packages/core/utils/src/validate/index.ts

```diff
@@ -0,0 +1,132 @@
+import { CurriedFunction1 } from 'lodash';
+import { isArray } from 'lodash/fp';
+
+import { getNonWritableAttributes } from '../content-types';
+import { pipeAsync } from '../async';
+
+import * as visitors from './visitors';
+import * as validators from './validators';
+import traverseEntity, { Data } from '../traverse-entity';
+
+import traversals from '../traverse/traversals';
+
+import { Model } from '../types';
+
+const { traverseQueryFilters, traverseQuerySort } = traversals;
+
+export interface Options {
+  auth?: unknown;
+}
+
+interface Validator {
+  (schema: Model): CurriedFunction1<Data, Promise<Data>>;
+}
+export interface ValidateFunc {
+  (data: unknown, schema: Model, options?: Options): Promise<void>;
+}
+
+const createContentAPIValidators = () => {
+  const validateInput: ValidateFunc = async (data: unknown, schema: Model, { auth } = {}) => {
+    if (isArray(data)) {
+      await Promise.all(data.map((entry) => validateInput(entry, schema, { auth })));
+      return;
+    }
+
+    const nonWritableAttributes = getNonWritableAttributes(schema);
+
+    const transforms = [
+      // Remove non writable attributes
+      traverseEntity(visitors.throwRestrictedFields(nonWritableAttributes), { schema }),
+    ];
+
+    if (auth) {
+      // Remove restricted relations
+      transforms.push(traverseEntity(visitors.throwRestrictedRelations(auth), { schema }));
+    }
+
+    // Apply validators from registry if exists
+    strapi.validators
+      .get('content-api.input')
+      .forEach((validator: Validator) => transforms.push(validator(schema)));
+
+    pipeAsync(...transforms)(data as Data);
+  };
+
+  const validateQuery = async (
+    query: Record<string, unknown>,
+    schema: Model,
+    { auth }: Options = {}
+  ) => {
+    if (!schema) {
+      throw new Error('Missing schema in validateQuery');
+    }
+    const { filters, sort, fields } = query;
+
+    if (filters) {
+      await validateFilters(filters, schema, { auth });
+    }
+
+    if (sort) {
+      await validateSort(sort, schema, { auth });
+    }
+
+    if (fields) {
+      await validateFields(fields, schema);
+    }
+
+    // TODO: validate populate
```

Comment:
Is this intentional?

Reply:
Yes, it was decided populate was out of scope for the time being, and I didn't want it to look "complete"